File: android/src/main/java/com/youversion/reactnativesdk/RNYouVersionPlatformModule.kt
package com.youversion.reactnativesdk

import com.youversion.platform.core.YouVersionPlatformConfiguration
import com.youversion.reactnativesdk.api.BibleReferenceRecord
import com.youversion.reactnativesdk.api.YVPBibleApi
import com.youversion.reactnativesdk.api.YVPHighlightsApi
import com.youversion.reactnativesdk.api.YVPLanguagesApi
import com.youversion.reactnativesdk.api.YVPVotdApi
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.functions.Coroutine

class RNYouVersionPlatformModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("RNYouVersionPlatform")

        Function("configure") { appKey: String ->
            val context = appContext.reactContext
                ?: throw IllegalStateException("ReactContext is not available yet")

            YouVersionPlatformConfiguration.configure(
                context = context,
                appKey = appKey,
            )
        }

        Function("setApiHost") { apiHost: String ->
            YouVersionPlatformConfiguration.setApiHost(apiHost)
        }
        
        Function("getAccessToken") {
            YouVersionPlatformConfiguration.accessToken
        }

        AsyncFunction("verseOfTheDay") Coroutine { dayOfYear: Int ->
            return@Coroutine YVPVotdApi.verseOfTheDay(
                dayOfYear = dayOfYear
            )
        }

        AsyncFunction("languages") Coroutine { country: String? ->
            return@Coroutine YVPLanguagesApi.languages(
                country = country
            )
        }

        AsyncFunction("versions") Coroutine { languageTag: String? ->
            return@Coroutine YVPBibleApi.versions(
                languageTag = languageTag
            )
        }

        AsyncFunction("version") Coroutine { versionId: Int ->
            return@Coroutine YVPBibleApi.version(
                versionId = versionId
            )
        }

        AsyncFunction("chapter") Coroutine { bibleReference: BibleReferenceRecord ->
            val context = appContext.reactContext
                ?: throw IllegalStateException("ReactContext is not available yet")

            return@Coroutine YVPBibleApi.chapter(
                bibleReference = bibleReference,
                context = context
            )
        }

        AsyncFunction("createHighlight") Coroutine { bibleId: Int, passageId: String, color: String ->
            return@Coroutine YVPHighlightsApi.createHighlight(
                bibleId = bibleId,
                passageId = passageId,
                color = color
            )
        }

        AsyncFunction("getHighlights") Coroutine { bibleId: Int, passageId: String ->
            return@Coroutine YVPHighlightsApi.getHighlights(
                bibleId = bibleId,
                passageId = passageId,
            )
        }

        AsyncFunction("updateHighlight") Coroutine { bibleId: Int, passageId: String, color: String ->
            return@Coroutine YVPHighlightsApi.updateHighlight(
                bibleId = bibleId,
                passageId = passageId,
                color = color
            )
        }

        AsyncFunction("deleteHighlight") Coroutine { bibleId: Int, passageId: String ->
            return@Coroutine YVPHighlightsApi.deleteHighlight(
                bibleId = bibleId,
                passageId = passageId
            )
        }
    }
}

File: ios/RNYouVersionPlatformModule.swift
import ExpoModulesCore

public class RNYouVersionPlatformModule: Module {
    public func definition() -> ModuleDefinition {
        Name("RNYouVersionPlatform")

        AsyncFunction("configure") { (appKey: String) async -> Void in
            await YVPlatform.configure(appKey: appKey)
        }
        
        AsyncFunction("setApiHost") { (apiHost: String) async -> Void in
            YVPlatform.setApiHost(apiHost: apiHost)
        }
        
        Function("getAccessToken") { () -> String? in
            return YVPlatform.getAccessToken()
        }

        AsyncFunction("signIn") { (
            permissions: [String],
            promise: Promise
        ) in
            YVPAuthAPI.signIn(
                permissions: permissions,
                promise: promise
            )
        }
        
        AsyncFunction("signOut") { () async -> Void in
            await YVPAuthAPI.signOut()
        }
        
        AsyncFunction("userInfo") { (
            promise: Promise
        ) in
            YVPAuthAPI.userInfo(promise: promise)
        }
        
        AsyncFunction("verseOfTheDay") { (
            dayOfYear: Int,
            promise: Promise
        ) in
            YVPVotdAPI.verseOfTheDay(dayOfYear: dayOfYear, promise: promise)
        }
        
        AsyncFunction("createHighlight") { (bibleId: Int, passageId: String, color: String, promise: Promise) in
            YVPHighlightsAPI.createHighlight(bibleId: bibleId, passageId: passageId, color: color, promise: promise)
        }
        
        AsyncFunction("getHighlights") { (bibleId: Int, passageId: String, promise: Promise) in
            YVPHighlightsAPI.getHighlights(bibleId: bibleId, passageId: passageId, promise: promise)
        }
        
        AsyncFunction("updateHighlight") { (bibleId: Int, passageId: String, color: String, promise: Promise) in
            YVPHighlightsAPI.updateHighlight(bibleId: bibleId, passageId: passageId, color: color, promise: promise)
        }
        
        AsyncFunction("deleteHighlight") { (bibleId: Int, passageId: String, promise: Promise) in
            YVPHighlightsAPI.deleteHighlight(bibleId: bibleId, passageId: passageId, promise: promise)
        }
        
        AsyncFunction("languages") { (country: String?, promise: Promise) in
            YVPLanguagesAPI.languages(country: country, promise: promise)
        }
        
        AsyncFunction("versions") { (languageTag: String?, promise: Promise) in
            YVPBibleAPI.versions(languageTag: languageTag, promise: promise)
        }
        
        AsyncFunction("version") { (versionId: Int, promise: Promise) in
            YVPBibleAPI.version(versionId: versionId, promise: promise)
        }
        
        AsyncFunction("chapter") { (bibleReference: BibleReferenceRecord, promise: Promise) in
            YVPBibleAPI.chapter(bibleReference: bibleReference, promise: promise)
        }
    }
}
